fairlearn/metrics/_group_metric_set.py

```diff
@@ -0,0 +1,169 @@
+        self._metrics = None
+
+    @property
+    def model_type(self):
+        """Return the type of model being described."""
+        return self._model_type
+
+    @model_type.setter
+    def model_type(self, value):
+        if value not in GroupMetricSet._allowed_model_types:
+            msg_format = "model_type '{0}' not in {1}"
+            msg = msg_format.format(value, sorted(list(GroupMetricSet._allowed_model_types)))
+            raise ValueError(msg)
+        self._model_type = value
+
+    @property
+    def y_true(self):
+        """Return the array of y_true values."""
+        return self._y_true
+
+    @y_true.setter
+    def y_true(self, value):
+        self._y_true = _convert_to_ndarray_1d(value)
+
+    @property
+    def y_pred(self):
+        """Return the array of y_pred values."""
+        return self._y_pred
+
+    @y_pred.setter
+    def y_pred(self, value):
+        self._y_pred = _convert_to_ndarray_1d(value)
+
+    @property
+    def groups(self):
+        """Return the array of group values."""
+        return self._groups
+
+    @groups.setter
+    def groups(self, value):
+        self._groups = _convert_to_ndarray_1d(value)
+
+    @property
+    def group_names(self):
+        """Return the group_names dictionary."""
+        return self._group_names
+
+    @group_names.setter
+    def group_names(self, value):
+        key_types = set(type(k) for k in value.keys())
+        if key_types != {int}:
+            raise ValueError("Keys not integers")
+        value_types = set(type(v) for v in value.values())
+        if value_types != {str}:
+            raise ValueError("Values not strings")
+
+        self._group_names = value
+
+    @property
+    def metrics(self):
+        """Return the GUID-GroupMetricResult dictionary of group metrics."""
+        return self._metrics
+
+    @metrics.setter
+    def metrics(self, value):
+        key_types = set(type(k) for k in value.keys())
+        if key_types != {str}:
+            raise ValueError("Keys not strings")
+        value_types = set(type(v) for v in value.values())
+        if value_types != {GroupMetricResult}:
+            raise ValueError("Values not GroupMetricResults")
+        self._metrics = value
+
+    def compute(self, y_true, y_pred, groups, model_type=BINARY_CLASSIFICATION, group_names=None):
+        """Compute the default metrics."""
+        self.y_true = y_true
+        self.y_pred = y_pred
+        self.groups = groups
+        self.model_type = model_type
+        if group_names is not None:
+            self.group_names = group_names
+
+        computed_metrics = {}
+        if self.model_type == GroupMetricSet.BINARY_CLASSIFICATION:
+            for k, f in GroupMetricSet.BINARY_CLASSIFICATION_METRICS.items():
+                computed_metrics[k] = f(self.y_true, self.y_pred, self.groups)
+        elif self.model_type == GroupMetricSet.REGRESSION:
+            for k, f in GroupMetricSet.REGRESSION_METRICS.items():
```

Comment:
so k and f are metric_name and metric_function? Can we name it that?

Reply:
Fair point. Have taken the opportunity to do a bit of other refactoring